Unify assignment of OwnerRRef result

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58753 Fix race condition in TP agent
* #58675 Make Future always extract DataPtrs in RPC tests
* #56863 Ensure async_execution works with CUDAFuture
* #57355 Avoid re-doing CUDA stream sync in OwnerRRef
* #58428 Make TP agent use streams from Future when sending response
* #58674 Set and propagate devices in RRef completion future
* #58427 Set streams when invoking UDFs
* #58426 Create CUDA-aware futures in RequestCallback
* #58425 Provide pre-extracted DataPtrs when completing a Future with a Message
* #58424 Allow Future::then to return pre-extracted DataPtrs
* #58423 Always use intrusive_ptr for Message (2 out of 2)
* #58422 Always use intrusive_ptr for Message (1 out of 2)
* #58421 Prevent using anything other than intrusive_ptr for Future
* #58420 Migrate remaining shared_ptr<Future> to intrusive_ptr
* #57861 Make remaining autograd methods return futures
* #57860 Make remaining RRef methods return futures
* #57859 Unify fetching RRefs
* #57858 Deduplicate Python object serialization
* #57857 Simplify process(Script|Python)(Remote)?Call
* **#57856 Unify assignment of OwnerRRef result**
* #57855 Make processPythonExecution return a future
* #57854 Remove getScriptRemoteCallType
* #57853 Simplify OwnerRRef completion
* #57852 Unify async execution for JIT functions
* #57851 Unify invoking JIT functions
* #57850 Unify invoking JIT operands
* #57849 Make some methods of RequestCallback return void instead of bool
* #57848 Centralize setting messageId in RequestCallback
* #57847 Make RequestCallback collect Futures from methods, rather than providing them
* #57846 Add helpers to manipulate futures
* #57638 Reduce overhead when Future invokes callbacks inline
* #57637 Introduce thenAsync method on Future

Thanks to Futures providing a "common language" between various steps, we can now deduplicate the creation of OwnerRRef, by having two different ways of creating the result (JIT and Python) but then connecting them to a single method that wraps and stores that result in an OwnerRRef.

Differential Revision: [D28253845](https://our.internmc.facebook.com/intern/diff/D28253845/)

lw committed:
Unify assignment of OwnerRRef result

Thanks to Futures providing a "common language" between various steps, we can now deduplicate the creation of OwnerRRef, by having two different ways of creating the result (JIT and Python) but then connecting them to a single method that wraps and stores that result in an OwnerRRef.

Differential Revision: [D28253845](https://our.internmc.facebook.com/intern/diff/D28253845/)

[ghstack-poisoned]

lw committed:
Update on "Unify assignment of OwnerRRef result"


Thanks to Futures providing a "common language" between various steps, we can now deduplicate the creation of OwnerRRef, by having two different ways of creating the result (JIT and Python) but then connecting them to a single method that wraps and stores that result in an OwnerRRef.

Differential Revision: [D28253845](https://our.internmc.facebook.com/intern/diff/D28253845/)

[ghstack-poisoned]

lw committed:
Update on "Unify assignment of OwnerRRef result"


Thanks to Futures providing a "common language" between various steps, we can now deduplicate the creation of OwnerRRef, by having two different ways of creating the result (JIT and Python) but then connecting them to a single method that wraps and stores that result in an OwnerRRef.

Differential Revision: [D28253845](https://our.internmc.facebook.com/intern/diff/D28253845/)

[ghstack-poisoned]

lw committed:
Update on "Unify assignment of OwnerRRef result"


Thanks to Futures providing a "common language" between various steps, we can now deduplicate the creation of OwnerRRef, by having two different ways of creating the result (JIT and Python) but then connecting them to a single method that wraps and stores that result in an OwnerRRef.

Differential Revision: [D28253845](https://our.internmc.facebook.com/intern/diff/D28253845/)

[ghstack-poisoned]

mrshenli reviewed (commented):
Test failures are real, but it might be because we have a better (more specific) type now?

lw committed:
Update on "Unify assignment of OwnerRRef result"


Thanks to Futures providing a "common language" between various steps, we can now deduplicate the creation of OwnerRRef, by having two different ways of creating the result (JIT and Python) but then connecting them to a single method that wraps and stores that result in an OwnerRRef.

Differential Revision: [D28253845](https://our.internmc.facebook.com/intern/diff/D28253845/)

[ghstack-poisoned]

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -4794,6 +4798,10 @@ def _test_rref_proxy_timeout(self, rref_proxy_api):
 
+        # FIXME We wait until the remote completed creating the OwnerRRef
+        # because there's currently a race if we shut down RPC before that.
+        slow_rref.to_here()
```
cc @rohan-varma 

Is the `_get_future` API sufficient for this purpose? What's the contract on that API? If user A share it with user B, will this API work on user B?

https://github.com/pytorch/pytorch/blob/4aa91871c0364279b2f3d04c1ef5594ef1e17f58/torch/csrc/distributed/rpc/init.cpp#L442-L452

lw replied:
I've used `slow_rref.to_here()` because it's being used for this same exact purpose a few lines above:
https://github.com/pytorch/pytorch/blob/4d7abdbdadd440cb4b8412f1e309cae14a687b49/torch/testing/_internal/distributed/rpc/rpc_test.py#L4772-L4773

lw committed:
Update on "Unify assignment of OwnerRRef result"


Thanks to Futures providing a "common language" between various steps, we can now deduplicate the creation of OwnerRRef, by having two different ways of creating the result (JIT and Python) but then connecting them to a single method that wraps and stores that result in an OwnerRRef.

Differential Revision: [D28253845](https://our.internmc.facebook.com/intern/diff/D28253845/)

[ghstack-poisoned]

lw committed:
Update on "Unify assignment of OwnerRRef result"


Thanks to Futures providing a "common language" between various steps, we can now deduplicate the creation of OwnerRRef, by having two different ways of creating the result (JIT and Python) but then connecting them to a single method that wraps and stores that result in an OwnerRRef.

Differential Revision: [D28253845](https://our.internmc.facebook.com/intern/diff/D28253845/)

[ghstack-poisoned]

lw committed:
Update on "Unify assignment of OwnerRRef result"


Thanks to Futures providing a "common language" between various steps, we can now deduplicate the creation of OwnerRRef, by having two different ways of creating the result (JIT and Python) but then connecting them to a single method that wraps and stores that result in an OwnerRRef.

Differential Revision: [D28253845](https://our.internmc.facebook.com/intern/diff/D28253845/)

[ghstack-poisoned]
